pkg/sharky/shard.go

```diff
@@ -0,0 +1,188 @@
+	err error    // signal for end of operation
+}
+
+// read models the input to read operation (the output is an error)
+type read struct {
+	buf  []byte // variable size read buffer
+	slot uint32 // slot to read from
+	at   int    // within-chunk offset for partial reads
+}
+
+// shard models a shard writing to a file with periodic offsets due to fixed datasize
+type shard struct {
+	reads    chan read     // channel for reads
+	errc     chan error    // result for reads
+	writes   chan write    // channel for writes
+	index    uint8         // index of the shard
+	datasize int           // max size of blobs
+	file     sharkyFile    // the file handle the shard is writing data to
+	slots    *slots        // component keeping track of freed slots
+	quit     chan struct{} // channel to signal quitting
+}
+
+// forever loop processing
+func (sh *shard) process() {
+	var writes chan write
+	var slot uint32
+	free := sh.slots.out
+LOOP:
+	for {
+		select {
+		case op := <-sh.reads:
+			// prioritise read ops i.e., continue processing read ops (only) as long as any
+			// this will block any writes on this shard effectively making store-wide
+			// write op to use a differenct shard while this one is busy
+			for {
+				sh.errc <- sh.read(op)
+				select {
+				case op = <-sh.reads:
+				default:
+					continue LOOP
+				}
+			}
+
+			// only enabled if there is a free slot previously popped
+		case op := <-writes:
+			op.res <- sh.write(op.buf, slot)
+			free = sh.slots.out // reenable popping a free slot next time we can write
+			writes = nil        // disable popping a write operation until there is a free slot
+
+			// pop a free slot
+		case slot = <-free:
+			// only if there is one can we pop a chunk to write otherwise keep back pressure on writes
+			// effectively enforcing another shard to be chosen
+			writes = sh.writes // enable popping a write operation
+			free = nil         // disabling getting a new slot until a write is actually done
+
+		case <-sh.quit:
+			// this condition checks if an slot is in limbo (popped but not used for write op)
+			if writes != nil {
+				sh.slots.wg.Add(1) // Done after the slots process pops from slots.in
+				go func() {
+					sh.slots.in <- slot
+				}()
+			}
+			return
+		}
+	}
+}
+
+// close closes the shard:
+// wait for pending operations to finish then saves free slots and blobs on disk
+func (sh *shard) close() error {
+	sh.slots.wg.Wait()
+	if err := sh.slots.save(); err != nil {
+		return err
+	}
+	if err := sh.slots.file.Close(); err != nil {
+		return err
+	}
+	return sh.file.Close()
+}
+
+// offset calculates the offset from the slot
+// this is possible since all blobs are of fixed size
+func (sh *shard) offset(slot uint32) int64 {
+	return int64(slot) * int64(sh.datasize)
+}
+
+// read reads loc.Length bytes to the buffer from the blob slot loc.Slot
+func (sh *shard) read(r read) error {
+	_, err := sh.file.ReadAt(r.buf, sh.offset(r.slot)+int64(r.at))
```

Comment:
let's remove `r.at`. it is not used anywhere, always zero, not plugged into any high-level API and far away from any point of integration.